### ISSUE TITLE
Results saving and reporting (Jira XRay)

### ISSUE THREAD

**Issue by mjftw** [Witekio/pluma-automation#184]
**Is your feature request related to a problem? Please describe.**
* Reporting to Jira XRay
* Save results to file to report later

**Describe the solution you'd like**
New code added to allow for something like the following design.
![Screenshot from 2021-01-18 17-30-54](https://user-images.githubusercontent.com/8429712/104947453-b1af3600-59b3-11eb-81c2-5db8c91685b9.png)

Xray and Slack are just down as examples, and do not need to be in this feature but can be later work.

To report saved results we'd probably use a new pluma subcommand like: `pluma report --json results.json email`

** Success criteria **
* Results can be saved to disk
* Results can be read from disk and passed to a reporter, allowing reporting to happen on a different `pluma` run.

**Comment by prototypicalpro** [comment]
I can take a look at implementing the reporting half of this issue this week if that sounds good to you @mjftw?

**Comment by prototypicalpro** [comment]
I noticed the PlanRunner XRay integration has support for pushing real-time updates and metrics, which are processed through hooks called during the test process:
```python
class JiraReporter(ResultConsumer):
    def on_plan_start(self, ...):
        ...

    def on_plan_end(self, ...):
        ...

    def on_session_start(self, ...):
        ...

    def on_session_end(self, ...):
        ...

    def on_test_start(self, ...):
        ...

    def on_metric_emited(self, ...):
        ...

    def on_result_emited(self, ...):
        ...

    def on_test_end(self, ...):
        ...
```
Are these features (real-time updates and metrics) important to the pluma XRay integration? It seems like a large part of the PlanRunner XRay code.

In addition, if we would like to support a feature such as real-time updates we may also want to consider adding a hook-based extension system such as [pluggy](https://pypi.org/project/pluggy/) (used by pytest), as a real-time update plugin will need to define functionality in the configuration, commandline, and every step of the test running process. This change would likely start with a large refactor of the `TestRunner` and `TestController` classes (moving their functionality into an event system), but would later allow modules to automatically register functionality to any stage in the testing process without the need for additional piping. I'd be happy to provide a more fleshed out example if that is a direction we'd like to pursue.

**Comment by prototypicalpro** [comment]
Update: I spoke with @Pixep and he strongly suggested including real-time progress reporting. Regarding pluggy, he agreed that it is an ideal tool to handle the one-to-many relationship with test progress reporting, but suggested starting with a simple POC XRay implementation without pluggy. We'd both be interested in any thoughts you have about integrating pluggy (or another system) into pluma.

**Comment by mjftw** [comment]
> Update: I spoke with @Pixep and he strongly suggested including real-time progress reporting. Regarding pluggy, he agreed that it is an ideal tool to handle the one-to-many relationship with test progress reporting, but suggested starting with a simple POC XRay implementation without pluggy first. We'd both be interested in any thoughts you have @mjftw about integrating pluggy (or another system) into pluma.

I agree about having it in there. It gives a much nicer user experience, and that's where all the perceived value of Pluma will come from, so this is a big win!
Pluggy sounds like a very sensible choice 👍🏼 But yeah let's take things slow and add pluggy after.

TL;DR - What @Pixep said!

Pluggy would actually be a nice way of handling our own user tests plugin management.
Thoughts @Pixep?

**Comment by Pixep** [comment]
> Pluggy would actually be a nice way of handling our own user tests plugin management.
Thoughts @Pixep?

Yeah that will be a great way to allowing plugging into some features directly, agree. I don't think it will fit all situations, but a good one to keep in ming.
